articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,315 @@
+
+Previously, the Fully Qualified Name (FQN) of a node was used directly by a selected security directory lookup strategy to load the necessary key material.
+However, now that Participants map to contexts and not nodes, such a direct mapping of node FQN to security artifacts is no longer appropriate.
+
+### Contexts
+
+With the advent of ROS 2, multiple nodes may now be composed into one process for improved performance.
+Previously however, each node would retain it's one to one mapping to a separate middleware ``Participant``.
+Given the non-negligible overhead incurred of multiple ``Participant``s per process, a change was introduced to map a single ``Participant`` to a context, and allow for multiple nodes to share that context.
+
+Based on the DDS Security specification v1.1, a ``Participant`` can only utilise a single security identity; consequently the access control permissions applicable to every node mapped to a given context must be consolidated and combined into a single set of security artifacts.
+As such, additional tooling and extensions to SROS 2 are necessary to support this new paradigm.
+
+## Keystore
+
+With the additional structure of contexts, it’s perhaps best to take the opportunity to restructure the keystore layout as well.
+Rather than a flat directory of namespaced node security directories, we can push all such security directories down into a designated `contexts` sub-folder.
+Similarly, private and public keystore materials can also be pushed down into their own respective sub-folders within the root keystore directory.
+This is reminiscent of the pattern used earlier Keymint [1].
+
+
+```
+$ tree keystore/
+keystore
+├── contexts
+│   └── ...
+│       └── ...
+├── private
+│   ├── ca.csr.pem
+│   └── ca.key.pem
+└── public
+    ├── ca.cert.pem
+    ├── identity_ca.cert.pem -> ca.cert.pem
+    └── permissions_ca.cert.pem -> ca.cert.pem
+```
+
+
+### ``public``
+
+The ``public`` directory contains anything permissible as public, such as public certificates for the identity or permissions certificate authorities.
+As such, this can be given read access to all executables.
+Note that in the default case, both the `identity_ca` and `permissions_ca` points to the same CA certificate.
+
+### ``private``
+
+The ``private`` directory contains anything permissable as private, such as private key material for aforementioned certificate authorities.
+This directory should be redacted before deploying the keystore onto the target device/robot.
+
+### ``contexts``
+
+The ``contexts`` directory contains the security artifacts associated with individual contexts, and thus node directories are no longer relevant.
+Similar to node directories however, the `contexts` folder may still recursively nest sub-paths for organizing separate contexts.
+The `ROS_SECURITY_ROOT_DIRECTORY` environment variable should by convention point to this directory.
```

Comment:
there is something that doesnt feel right about this.
as this is the "ROOT_DIRECTORY" it seems like I should be able to just ship that to my robot and it would "just work".
But files in the subfolders of `contexts` symlink to files outside of the "ROOT_DIRECTORY".

Also given that all the sros2 commands use the keystore directory as argument, it seems weird to say that the "root" is actually a subfolder of that.

I don't really remember, what was the rationale for not keeping `ROS_SECURITY_ROOT_DIRECTORY` pointing to the keystore directory ? to not append "contexts" arbitrarily in RCL ?
If it still makes sense to keep if pointing to `contexts`, could we change the env var name to something without "root" in it ? like `ROS_SECURITY_CONTEXTS_DIRECTORY` or something to that effect?

Reply:
> But files in the subfolders of contexts symlink to files outside of the "ROOT_DIRECTORY".
> Also given that all the sros2 commands use the keystore directory as argument, it seems weird to say that the "root" is actually a subfolder of that.

That's why I suggested that RCL should append the `contexts` sub folder too the root when knowing to look for context security artifacts, so that the root dir would span over any needed `public`, `contexts`, etc, directories. Given we are already hard coding conventions for artifact file names, it doesn't seem much of a stretch to hardcode conventions in the root directory structure.

> If it still makes sense to keep if pointing to contexts, could we change the env var name to something without "root" in it ? like ROS_SECURITY_CONTEXTS_DIRECTORY or something to that effect?

That could work, but I think reverting/correcting the ROOT env to point to same root keystore dir the the sros2 cli expects would help maintain consistency in user experience, as well as future proof the use of that ENV when we allow for setting more advance URI strings, other than just file system paths, like if using a TPM or remote keybase.

I left such comments on https://github.com/ros2/rcl/pull/515 , but it looks like github swallowed my remarks when the PR was rebased (losing review comments like that is annoying). :<

Reply:
> I left such comments on ros2/rcl#515 , but it looks like github swallowed my remarks when the PR was rebased (losing review comments like that is annoying). :<

Sorry if I lost a comment, but I needed to rebase.
Pointing to the `keystore` and not to the `contexts` directory sounds good to me.

I think that both `rcl` and `test_security` need to be updated.

Let me know if you're planning to send a PR, if not I will try to do it on Monday.

Reply:
> Sorry if I lost a comment, but I needed to rebase.

Don't worry, not faulting you, I blame github.
Just wasn't sure why comments are preserved differently than reviews.

See 09e5c9e for what I was first envisioning, and let me know if you see an issue.

Reply:
> See 09e5c9e for what I was first envisioning, and let me know if you see an issue.

SGTM!

Reply:
Fixed in https://github.com/ros2/rcl/pull/607.